compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -2874,6 +2888,10 @@ impl<'tcx> Ty<'tcx> {
     /// This is mostly useful for optimizations, as these are the types
     /// on which we can replace cloning with dereferencing.
     pub fn is_trivially_pure_clone_copy(self) -> bool {
+        if self.is_scalable_simd() {
```

Comment:
Why is this true?

Reply:
Actually, for the record, this is unsound. For a type to implement `Copy`, it *must* implement `Sized`, since `Sized` is one of `Copy`'s supertraits.

Reply:
Yes, that's one of the problems we have with these types. We need them to be `Copy` but we can't implement the trait due to the bounds. These types should be trivially copyable as they are just a register and (for SVE anyway) can be copied with a `mov` instruction. It was a discussion with @Amanieu a while ago that lead to this approach, I'm definitely open to alternative suggestions as to how this could be done though.

Reply:
This should probably be moved to `is_copy_modulo_regions`. Possibly further up, since it might depend on what the callers of that function are checking.

Reply:
`is_copy_modulo_regions` must agree with the implementation of `Copy`. As I said before, these types *cannot* implement `Copy` soundly while also being unsized.

Reply:
Then perhaps this exception should be moved up to the callers of is_copy_modulo_regions where we can get a better idea of what special handling is needed.

Reply:
Given that these types also have the requirements to only be used in registers, I would expect that trying to use too many if these variables at the same time is likely also not supported 🤔 Looking at the RFC I don't fully understand the requirements for these types, both the questions by RalfJ, e.g. https://github.com/rust-lang/rfcs/pull/3268/files#r1500263066,  but also why is it ok to reference them if they may only be stored in registers?

How does LLVM model `svfloat64_t` types? I would expect them to encounter many of the same issues?

Reply:
given that references to these simd registers are apparently ok and you also want their scope to be limited, could something like teh following work (where these simd registers are extern types with special behavior during codegen)
```rust
// An invariant token,similar to `scoped_threads` in std
#[derive(Copy, Clone)
struct Scope<'scope>(Phantomdata<*mut &'scope ()>);
fn sve_scope<F: for<'scope> Fn(Scope<'scope>) -> R>(f: F) -> R;

extern {
    type svbool<'scope>;
}

impl<'scope> Scope<'scope> {
    fn whilelt_b64(self, ..) -> &'scope mut svbool<'scope>;
    // ...
}
```

Given that I don't fully understand the constraints here there may definitely be issues with this approach